tests/testlib/s2n_ktls_test_utils.c

```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "testlib/s2n_ktls_test_utils.h"
+
+/* These MOCK_IO* macros set errno before returning an error. These macros are
+ * mainly used for IO related operations (stuffer writes, NULL arguments,
+ * invalid length). Other sanity checks are guarded by the POSIX* macros.
+ *
+ * EINVAL will always be treated as a fatal errno code so its sufficient to
+ * use that as the generic errno code in the guard macros. */
+#define MOCK_IO_ENSURE(x)   \
```

Comment:
It can be dangerous to try to use a different ENSURE/GUARD than developers are used to. I would suggest using the standard macros, but having a wrapper method that handles translating failures / errors into what you need them to be.

Reply:
https://github.com/aws/s2n-tls/pull/4109/commits/f24d55ff425ed89f73d6ca13b33ea6511099064a

simplified since we only need to handle EAGAIN when reading/writing to the data_buffer